scheduler/src/cook/mesos/api.clj

```diff
@@ -809,30 +815,58 @@
        :job/uuid))
 
 (defn retrieve-jobs
+  "Returns a tuple that either has the shape:
+
+    [true {::jobs ...
+           ::jobs-requested ...
+           ::instances-requested ...}]
+
+  or:
+
+    [false {::error ...}]
+
+  Given a collection of job and/or instance uuids, attempts to return the
+  corresponding set of existing job uuids. By default (or if the 'partial'
+  query parameter is false), the function returns an ::error if any of the
+  provided job or instance uuids cannot be found. If 'partial' is true, the
+  function will return the subset of job uuids that were found, assuming at
+  least one was found. This 'partial' flag allows a client to query a
+  particular cook cluster for a set of uuids, where some of them may not belong
+  to that cluster, and get back the data for those that do match."
   [conn ctx]
   (let [jobs (get-in ctx [:request :query-params :job])
-        instances (get-in ctx [:request :query-params :instance])]
+        instances (get-in ctx [:request :query-params :instance])
+        allow-partial-results (get-in ctx [:request :query-params :partial])]
     (let [instance-uuid->job-uuid #(instance-uuid->job-uuid (d/db conn) %)
           instance-jobs (mapv instance-uuid->job-uuid instances)
-          used? (partial job-exists? (db conn))]
+          exists? #(job-exists? (db conn) %)]
       (cond
-        (and (every? used? jobs)
-             (every? (complement nil?) instance-jobs))
+        (and (every? exists? jobs)
```

Comment:
nitpick: do we care about potential multiple calls to `exists?` over the execution of this function. Do we want to filter the jobs eagerly and work on the lengths of the collections in the conditionals?